library/Require/Transform.hs

```diff
@@ -1,78 +1,110 @@
 {-# LANGUAGE FlexibleContexts #-}
 module Require.Transform where
 
+import Control.Category ((>>>))
+import Control.Monad.Except
+import Control.Monad.Writer.Strict
+import Data.DList (DList)
 import qualified Data.Text as Text
 import Relude
+import Require.Error (Error(..))
 import qualified Require.File as File
 import qualified Require.Parser as Parser
 import Require.Types
 
 
-type LineTagPrepend = File.LineTag -> Text -> Text
+-- | The monad stack used during transformation:
+--
+-- * @'StateT' 'TransformState'@ to keep track of whether to render the next
+--   line tag, the module's name etc.
+-- * @'WriterT' ('DList' 'Text')@ to collect the output lines. Instead of
+--   haskell's built-in list type we use 'DList' because of it's /O(1)/ append
+--   operation.
+-- * @'Either' 'Error'@ to return errors.
+type TransformM =
+  StateT TransformState
+    (WriterT (DList Text)
+      (Either Error))
+
 
 data TransformState = TransformState
-  { tstLineTagPrepend :: !LineTagPrepend
+  { tstLineTagOutput  :: File.LineTag -> TransformM ()
   , tstHostModule     :: !(Maybe ModuleName)
   , tstAutorequire    :: !(AutorequireMode File.Input)
   }
 
 
-renderLineTag :: File.LineTag -> Text
-renderLineTag (File.LineTag (File.Name fn) (File.LineNumber ln)) =
-  "{-# LINE " <> show ln <> " \"" <> fn <> "\" #-}\n"
-
+-- | Outputs a single line.
+output :: Text -> TransformM ()
+output = tell . pure
 
-prependLineTag :: LineTagPrepend
-prependLineTag = (<>) . renderLineTag
+-- | Outputs the pragma representation of the given line tag.
+renderLineTag :: File.LineTag -> TransformM ()
+renderLineTag (File.LineTag (File.Name fn) (File.LineNumber ln)) =
+  output $ "{-# LINE " <> show ln <> " \"" <> fn <> "\" #-}"
 
-ignoreLineTag :: LineTagPrepend
-ignoreLineTag = const id
+-- | Ignore the given line tag, specifically don't render it.
+ignoreLineTag :: File.LineTag -> TransformM ()
+ignoreLineTag = const (pure ())
 
 
-transform :: AutorequireMode File.Input -> File.Input -> Text
-transform autorequire input =
-  -- TODO:
-  --  * if the mapM overhead is too much maybe use a streaming library
-  --  * there is no need to concatenate the whole output in memory, a lazy text would be fine
-  --  * maybe we should check if tstAutorequired is set after processing
-  File.inputLines input
-    & mapM (process (pure Nothing))
-    & flip evalState initialState
-    & mconcat
+transform :: AutorequireMode File.Input -> File.Input -> Either Error [Text]
+transform autorequire =
+  File.inputLines
+    >>> traverse_ (process False)
+    >>> flip execStateT initialState
+    >>> (>>= checkDidAutorequire)
+    >>> execWriterT
+    >>> fmap toList
```

Comment:
I'd avoid using arrow composition here. I'd add another parameter to the function so it's not point-free and then chain the calls using `&`

Reply:
Would you prefer to keep the section `(>>= …)` or use `>>=` in the pipeline?

```hs
  & flip …
  & (>>= …)
  & execWriterT
```

vs

```hs
  & flip …
  >>= …
  & execWriterT
```

They both work as `>>=` is also `infixl 1`, I have no real preference either way.

Reply:
I'd use [`chainedTo` from relude](https://hackage.haskell.org/package/relude-0.6.0.0/docs/Relude-Monad.html#v:chainedTo)

Reply:
Oh yes, that is a nice function! Didn’t know about it